### ISSUE TITLE
M2#A128

### ISSUE THREAD

**Issue by fmussap** [da2k/curso-reactjs-ninja#61]
Fernando,

Se voce nao possuir nennhuma key, ou seja, nao ha markdown-editor gravado, ele apresenta um erro no ao ler o arquivo **files.js**.
Se voce ja possuir a chave markdwon-editor criada, mesmo que vazia, ai nao apresenta problema.

Para nao apresentar esse erro fiz o seguinte abaixo paraver antes se ha conteudo em files:

**files && Object.keys(files).map((fileId)  ...**

Nao sei se esta e a melhor solucao

@fdaciuk

**Comment by fdaciuk** [comment]
Boa @fmussap! Acho que eu não testei esse caso específico :)

Só um ponto: quando usar curto circuito (com `&&`), ele sempre vai te retornar um valor, então ou você atribui essa expressão a uma variável, ou usa `if` mesmo :)

**Comment by fmussap** [comment]
```
const Files = ({ files, handleOpenFile }) => (
  <div className='files-list-container'>
    <h2>Files</h2>
    <ul>
      {files && Object.keys(files).map((fileId) => (
        <li key={fileId}>
          <button onClick={handleOpenFile(fileId)}>{files[fileId].title}</button>
        </li>
      ))}
    </ul>
  </div>
)
```

So para entender melhor essa parte do &&, o que poderia dar de errado nesse caso. Eu pensei que ele so ia criar o `<li>` se **files** tivesse valor. 
Que possibildades de valor ele poderia retornar?

Obrigado de  novo

**Comment by fdaciuk** [comment]
Certo! Agora olhando o seu código eu vi onde você está testando o `files` xD
Na verdade, nem precisaria dessa validação, já que nós passamos o `this.state.files` para esse componente, e se o `this.state.files` não existir, ele vai pegar o valor inicial, que é um objeto :)

Se quiser garantir que esse valor vai chegar nesse componente como um objeto, você pode usar `defaultProps` pra isso, aí nem precisa da validação :)

Se quiser, pode usar até o _default parameters_ do ES6+, fazendo assim:

```jsx
const Files = ({ files = {}, handleOpenFile }) => (
```

Dessa forma, se `files` for `undefined`, ele vai receber um objeto e não vai quebrar seu `map` ;)

E sobre o problema de usar `&&`, nesse seu caso não é um problema porque você está usando como expressão mesmo, dentro do JSX. Só se você estivesse usando fora do JSX, você deveria usar `if` no lugar :)

**Comment by fdaciuk** [comment]
Tentando explicar melhor sobre o `&&`:

```js
files && Object.keys(files)
```

Isso é uma expressão, e vai te retornar um valor. Um valor você normalmente usa para atribuir à uma variável, retornar de uma função, usar como expressão mesmo, saca?

Se esse código não estivesse no meio do JSX, sendo usado como expressão que retorna um valor para o JSX, então você teria que usar o `if`, para deixar o código mais "semântico":

```js
if (!files) {
  return 
}

return (
  <ul>
    {Object.keys(files).map(...)}
  </ul>
)
```

Algo assim, só pra exemplificar :)

**Comment by fmussap** [comment]
```
const Files = ({ files = {}, handleOpenFile }) => (
  <div className='files-list-container'>
    <h2>Files</h2>
    <ul>
      {Object.keys(files).map((fileId) => (
        <li key={fileId}>
          <button onClick={handleOpenFile(fileId)}>{files[fileId].title}</button>
        </li>
      ))}
    </ul>
  </div>
)
```

Fernando,

Alterei dessa forma e tentei com o defaultprops tambem e deu na mesma...

```
Uncaught TypeError: Cannot convert undefined or null to object
    at Function.keys (<anonymous>)
    at new Files (files.js:9)
```

Poderia dar uma olhada direto no repositorio?

https://github.com/fmussap/my-markdown-editor

**Comment by fdaciuk** [comment]
Oi Fábio! Vou rodar aqui pra ver o que pode estar acontecendo :D

**Comment by fdaciuk** [comment]
Oi @fmussap! Na verdade o problema acontece porque estamos tentando pegar os arquivos do localstorage, e assumindo que eles são um objeto. Quando ainda não temos a entrada, o `localStorage.getItem('markdown-editor')` retorna `null`, e o `JSON.parse` não consegue fazer o parse.

O certo seria garantir que isso vai ser, pelo menos, um objeto vazio, fazendo com que, se retornar null, passamos um objeto vazio, dessa forma:

```js
const files = JSON.parse(localStorage.getItem('markdown-editor')) || {}
```

Isso vai resolver :)

**Comment by fmussap** [comment]
Obrigado Fernando,

Funcionou perfeitamente!

Nesse caso utilizar o default parameters do ES6+ no arquivo files.js nao funciona 100%, certo?

**Comment by fdaciuk** [comment]
Funciona! Mas o `default parameters` só usa o valor passado se a variável for `undefined`. Como o `getItem` do JSON retorna `null`, ele recebe esse valor `null` :)

**Comment by fdaciuk** [comment]
Qualquer dúvida, só avisar! :D
